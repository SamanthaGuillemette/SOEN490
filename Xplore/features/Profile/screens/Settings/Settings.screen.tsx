import { StatusBar } from "expo-status-bar";
import { TouchableOpacity, View } from "react-native";
import { ScrollView } from "react-native";
import { NavigationProp } from "@react-navigation/native";
import { SafeAreaView } from "react-native-safe-area-context";
import { Avatar, Icon, ShadowView } from "../../../../components";
import { EditButton } from "../../components";
import { useEffect, useState } from "react";
import * as ImagePicker from "expo-image-picker";
import { useAuth, useThemeColor } from "../../../../hooks";
import api from "../../../../services/appwrite/api";
import { BUCKET_PROFILE_PIC, COLLECTION_ID_USERS } from "@env";
import { uploadImageToServer } from "../../../../utils";
import { useMutation, useQuery, useQueryClient } from "react-query";
import styles from "./Settings.styles";
import { Query } from "appwrite";

interface SettingsProps {
  navigation: NavigationProp<any>;
}

const Settings = (props: SettingsProps) => {
  const { navigation } = props;
  const { sessionToken } = useAuth();
  const statusBarBg = useThemeColor("background");
  const [userId] = useState<string>(sessionToken?.userId!);
  const [userDocumentId, setUserDocumentId] = useState<string>();
  const [profilePictureId, setProfilePictureId] = useState<string>();
  const [localProfilePic, setLocalProfilePic] = useState<URL>();

  // TODO: Move this function to a separate file later
  const getUserInfo = async (userID: string) => {
    const response = await api.listDocuments(COLLECTION_ID_USERS, [
      Query.equal("userID", userID),
    ]);
    return response?.documents[0];
  };

  // TODO: Move this function to a separate file later
  const updateUserInfo = async (userID: string, data: any) => {
    const response = await api.updateDocument(
      COLLECTION_ID_USERS,
      userID,
      data
    );
    return response;
  };

  const { data: userObject } = useQuery("user", () => getUserInfo(userId));

  useEffect(() => {
    console.log(
      "====> INSIDE Settings.screen - userObject: ",
      JSON.stringify(userObject, null, 2)
    );
    setUserDocumentId(userObject?.$id); // This is the actual document Id - DIFFERENT from the 'userId'
    setProfilePictureId(userObject?.profilePicture);
  }, [userObject]);

  const { data: profilePicture } = useQuery(
    "profile picture",
    () =>
      api.getFilePreview(
        BUCKET_PROFILE_PIC,
        profilePictureId ?? "63edd6618559f6420892"
      ),
    {
      // This query will only run if "profilePictureId" is valid
      enabled: !!profilePictureId,
    }
  );

  useEffect(() => {
    setLocalProfilePic(profilePicture);
  }, [profilePicture]);

  const queryClient = useQueryClient();
  const mutation = useMutation(
    (updatedProfieImageId: string) => {
      return updateUserInfo(userDocumentId!, {
        profilePicture: updatedProfieImageId,
      });
    },
    {
      // On success of this action, invalidate & refetch the "user" query & "profile picture" query
      onSuccess: () => {
        queryClient.invalidateQueries("user");
        queryClient.invalidateQueries("profile picture");
      },
    }
  );

  const handleUploadImage = async () => {
    // No permissions request is necessary for launching the image gallery
    let pickedImage = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [3, 3],
      quality: 1,
    });

    if (!pickedImage.cancelled) {
      setLocalProfilePic(pickedImage.uri as any); // Set local image so that the picture is displayed right away.

      const uploadedImageResult: any = await uploadImageToServer(
        pickedImage.uri,
        BUCKET_PROFILE_PIC
      );
      // console.log(
      //   "======> uploadedImageResult",
      //   JSON.stringify(uploadedImageResult, null, 2)
      // );

      // Need to check "userDocumentId" -> use inside the mutate function
      if (userDocumentId) {
        mutation.mutate(uploadedImageResult?.$id);
      }
    }
  };

  return (
    <>
      <StatusBar style="auto" backgroundColor={statusBarBg} />
      <SafeAreaView
        edges={["top", "left", "right"]}
        style={styles.safeAreaStyle}
      >
        <ScrollView>
          <View style={styles.backButton}>
            <TouchableOpacity
              onPress={() => {
                navigation.goBack();
              }}
            >
              <Icon name="chevron-left" color="primary" />
            </TouchableOpacity>
          </View>

          <View style={styles.avatarContainer}>
            <Avatar size={135} name="user avatar" imageURL={localProfilePic} />
            <TouchableOpacity
              style={styles.editAvatarButton}
              onPress={handleUploadImage}
            >
              <ShadowView
                backgroundColor="white"
                style={styles.editAvatarButtonShadow}
              >
                <Icon name="edit" size="small" />
              </ShadowView>
            </TouchableOpacity>
          </View>

          <View style={styles.settingButtonsContainer}>
            <EditButton
              iconName="user"
              label="Edit Profile"
              onPress={() => alert("pressed!")}
            />
            <EditButton
              iconName="lock"
              label="Change Password"
              onPress={() => navigation.navigate("UpdatePassword")}
            />
            <EditButton
              iconName="smile"
              label="Update Interests"
              onPress={() => {
                navigation.navigate("TopicSelection");
              }}
            />
          </View>
        </ScrollView>
      </SafeAreaView>
    </>
  );
};

export default Settings;
