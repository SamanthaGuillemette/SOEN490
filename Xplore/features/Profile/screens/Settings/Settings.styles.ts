import { StyleSheet } from "react-native";
import { deviceScreenWidth } from "../../../../constants";

export default StyleSheet.create({
  safeAreaStyle: {
    flex: 1,
  },
  avatarContainer: {
    alignItems: "center",
  },
  editAvatarButton: {
    position: "absolute",
    bottom: 10,
    right: deviceScreenWidth / 2 - 60,
    zIndex: 1,
  },
  editAvatarButtonShadow: {
    padding: 8,
    borderRadius: 50,
  },
  settingButtonsContainer: {
    marginTop: 75,
    paddingHorizontal: 30,
  },
});
