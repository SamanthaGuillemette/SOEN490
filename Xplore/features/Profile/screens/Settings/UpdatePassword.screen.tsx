import { useState } from "react";
import { ScrollView } from "react-native";
import { PrimaryButton, Text, TextInput, View } from "../../../../components";
import { useThemeColor } from "../../../../hooks";
import api from "../../../../services/appwrite/api";
import styles from "./UpdatePassword.styles";

const UpdatePassword = () => {
  const [password, setPassword] = useState<string>("");
  const [passwordReentered, setReenteredPassword] = useState<string>("");
  const [buttonLabel, setButtonLabel] = useState<string>("SAVE");
  const errorColor = useThemeColor("error");

  const renderErrorText = () => {
    if (password !== passwordReentered) {
      return (
        <Text variant="body" style={{ color: errorColor }}>
          Re-entered password doesn't match
        </Text>
      );
    }
    return null;
  };

  const handleUpdatePassword = async () => {
    try {
      await api.updatePassword(password);
    } catch (error) {
      alert(error);
    }
    setTimeout(() => setButtonLabel("UPDATING..."), 300);
  };

  return (
    <ScrollView>
      <View style={styles.container}>
        <View style={styles.textItems}>
          <Text variant="h2">Change password</Text>
          <Text variant="body" color="bodyText" style={styles.subTitleText}>
            Enter a new password below
          </Text>
        </View>

        <TextInput
          placeHolder="New password"
          iconName="lock"
          secureTextEntry
          onChangeText={(pw: string) => setPassword(pw)}
        />
        <TextInput
          placeHolder="Re-enter password"
          iconName="lock"
          secureTextEntry
          onChangeText={(pw: string) => setReenteredPassword(pw)}
        />

        {renderErrorText()}

        <PrimaryButton
          label={buttonLabel}
          style={styles.primaryButton}
          onPress={handleUpdatePassword}
        />
      </View>
    </ScrollView>
  );
};

export default UpdatePassword;
