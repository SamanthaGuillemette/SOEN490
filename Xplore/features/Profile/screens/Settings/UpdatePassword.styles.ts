import { StyleSheet } from "react-native";

export default StyleSheet.create({
  container: {
    paddingTop: "25%",
    alignItems: "center",
    paddingHorizontal: "10%",
  },
  textItems: {
    marginVertical: "20%",
    alignItems: "center",
  },
  subTitleText: {
    textAlign: "center",
    paddingHorizontal: "10%",
    marginTop: 10,
  },
  primaryButton: {
    marginTop: "20%",
    marginBottom: 25,
  },
});
