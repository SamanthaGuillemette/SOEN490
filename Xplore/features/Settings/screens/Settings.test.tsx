import { render } from "@testing-library/react-native";
import Settings from "./Settings.screen";

describe("Settings />", () => {
  it("renders correctly", () => {
    const container = render(<Settings />);
    expect(container).toBeTruthy();
  });
