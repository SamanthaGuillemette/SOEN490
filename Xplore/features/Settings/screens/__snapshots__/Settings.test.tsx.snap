// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tesing for <Settings /> screen: matches snapshot 1`] = `null`;
